domain/user/service/service.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+
+	"github.com/juju/juju/core/user"
+	usererrors "github.com/juju/juju/domain/user/errors"
+)
+
+// State describes retrieval and persistence methods for upgrade info.
+type State interface {
+	// AddUser will add a new user to the database. If the user already exists
+	// an error that satisfies usererrors.AlreadyExists will be returned.
+	AddUser(context.Context, string, string, string, string) error
+
+	// GetUser will retrieve the user specified by name from the database where
+	// the user is active and has not been deleted. If the user does not exist
+	// or is deleted an error that satisfies usererrors.NotFound will be
+	// returned.
+	GetUser(context.Context, string) (user.User, error)
+	// RemoveUser marks the user as deleted. This obviates the ability of a user
+	// to function, but keeps the user retaining provenance, i.e. auditing.
+	RemoveUser(context.Context, string) error
+	// SetPassword removes any active activation keys and sets the user password
+	SetPassword(context.Context, string, user.UserPassword) error
+	// GenerateUserActivationKey makes a new activation key for the given user.
```

Comment:
This func needs to return the activation key.

This func also needs to state that it will remove any previously set password for the user.

It also needs to state that if the user authentication is currently disabled a specific error will be returned saying we can't generate activation keys for disabled users.

Reply:
We also need to fill out the contract on what happens if the user is not found.